### PR TITLE
chore: plugin required chains

### PR DESCRIPTION
**Pull request by zhouhanseng**
closes #

### CODE REVIEW

File: packages/plugin-infra/src/types.ts [hunk 1]
```diff
@@ -134,7 +135,7 @@ export namespace Plugin.Shared {
     }
     export interface Web3EnableRequirement {
         /** This flag indicates the plugin entry in the composition entry should be hidden if the current chain is invalid. */
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts:137]:
```suggestion
        /** Plugin can declare what chain it supports. When the current chain is not supported, the composition entry will be hidden. */
```

**Review by Jack-Works — APPROVED** [review]

File: packages/maskbook/src/components/InjectedComponents/PostDialog.tsx [hunk 2]
```diff
@@ -456,13 +456,15 @@ export function CharLimitIndicator({ value, max, ...props }: CircularProgressPro
...
     const result = useActivatedPluginsSNSAdaptor().map((plugin) =>
         Result.wrap(() => {
             const entry = plugin.CompositionDialogEntry
             const unstable = plugin.enableRequirement.target !== 'stable'
-            const requireChainValid = Boolean(plugin.enableRequirement.web3?.compositionEntryRequiresChainIDValid)
-            if (!entry || (!chainIdValid && requireChainValid)) return null
+            const chainIdValid =
+                !Boolean(plugin.enableRequirement.web3?.compositionEntryRequiredChains) ||
+                Boolean(plugin.enableRequirement.web3?.compositionEntryRequiredChains?.includes(chainId))
+            if (!entry || !chainIdValid) return null
```
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/PostDialog.tsx:467]: Does this change mean that if the user chose an unmatched chain, the plugin would not render?

File: packages/maskbook/src/components/InjectedComponents/ToolboxHint.tsx [hunk 3]
```diff
@@ -192,7 +194,7 @@ export function ToolboxHint(props: ToolboxHintProps) {
                 <Image src={ToolIconURLs.encryptedmsg.image} width={19} height={19} />
                 <Typography className={classes.text}>{ToolIconURLs.encryptedmsg.text}</Typography>
             </MenuItem>,
-            chainIdValid ? (
+            RedPacket_Base.enableRequirement.web3?.compositionEntryRequiredChains?.includes(chainId) ? (
```
Comment (guanbinrui) [inline packages/maskbook/src/components/InjectedComponents/ToolboxHint.tsx:197]: It's a bit verbose to add this condition. Does it make sense to make a hook for this?

File: packages/maskbook/src/components/InjectedComponents/PostDialog.tsx (hunk 2 above)
Comment (guanbinrui) [reply packages/maskbook/src/components/InjectedComponents/PostDialog.tsx:467]: Never mind. It's OK to not render the composition entry.

File: packages/plugin-infra/src/types.ts [hunk 4]
```diff
@@ -133,8 +134,8 @@ export namespace Plugin.Shared {
         web3?: Web3EnableRequirement
     }
     export interface Web3EnableRequirement {
-        /** This flag indicates the plugin entry in the composition entry should be hidden if the current chain is invalid. */
-        compositionEntryRequiresChainIDValid?: boolean
+        /** Plugin can declare what chain it supports. When the current chain is not supported, the composition entry will be hidden. */
+        compositionEntryRequiredChains?: ChainId[]
```
Comment (guanbinrui) [inline packages/plugin-infra/src/types.ts:138]: Highly request `postInspectorRequiredChains` too! And add a switch network panel when the chosen chain is not matched with the expected.

**Review by guanbinrui — APPROVED** [review]

File: packages/maskbook/src/components/InjectedComponents/PostDialog.tsx (hunk 2 above)
Comment (Jack-Works) [reply packages/maskbook/src/components/InjectedComponents/PostDialog.tsx:469]: Another choice is to render the composition entry but in a disabled chip with a tip "this plugin is not available in the current network". That might be less confusing.

File: packages/plugin-infra/src/types.ts (hunk 4 above)
Comment (Jack-Works) [reply packages/plugin-infra/src/types.ts:140]: In the original design, it did have this kind of API, but the problem is, some plugins are rendering their own "Chain not supported" UI, therefore I don't know how to handle this case.

Comment (Jack-Works) [reply packages/plugin-infra/src/types.ts:140]: And another problem is, the current abstraction level doesn't support detecting if the plugin needs to render without actually render the content. I'll add two new APIs later that might resolve this problem.

**Review by Jack-Works — DISMISSED** [review]

File: packages/plugin-infra/src/manager/sns-adaptor.ts [hunk 5]
```diff
@@ -16,6 +17,17 @@ export function useActivatedPluginsSNSAdaptor() {
     return useSubscription(subscription)
 }
 
+export function usePluginRequiredChains() {
```
Comment (Jack-Works) [inline packages/plugin-infra/src/manager/sns-adaptor.ts:20]:
```suggestion
export function useActivatedPluginSNSAdaptorWithCompositionEntryRequirementMet() {
```

Comment (zhouhanseng) [reply packages/plugin-infra/src/manager/sns-adaptor.ts:20]: `usePluginRequiredChains()` just return a Plugin.ID => boolean mapping.

Comment (zhouhanseng) [reply packages/plugin-infra/src/manager/sns-adaptor.ts]: done

File: packages/plugin-infra/src/manager/sns-adaptor.ts [hunk 6]
```diff
@@ -16,6 +17,16 @@ export function useActivatedPluginsSNSAdaptor() {
     return useSubscription(subscription)
 }
 
+export function useActivatedPluginSNSAdaptorWithChainRequirementMet() {
+    const chainId = useChainId()
+    const plugins = useActivatedPluginsSNSAdaptor()
+    return plugins.reduce((acc: { [key: string]: boolean }, cur) => {
```
Comment (Jack-Works) [inline packages/plugin-infra/src/manager/sns-adaptor.ts:23]: Why not use filter?

File: packages/plugin-infra/src/types.ts [hunk 7]
```diff
@@ -137,8 +138,8 @@ export namespace Plugin.Shared {
         web3?: Web3EnableRequirement
     }
     export interface Web3EnableRequirement {
-        /** This flag indicates the plugin entry in the composition entry should be hidden if the current chain is invalid. */
-        compositionEntryRequiresChainIDValid?: boolean
+        /** Plugin can declare what chain it supports. When the current chain is not supported, the composition entry will be hidden. */
+        requiredChains?: ChainId[]
```
Comment (Jack-Works) [inline packages/plugin-infra/src/types.ts:142]: This name is misleading. It only hides the entrance for some cases but it didn't reflected that in the name.

File: packages/plugin-infra/src/manager/sns-adaptor.ts (hunk 6 above)
Comment (zhouhanseng) [reply packages/plugin-infra/src/manager/sns-adaptor.ts:23]: I just need a `{ [plugin.ID]: boolean }` mappings to indicate does each plugin support the current chain.

**Review by Jack-Works — DISMISSED** [review]

File: packages/web3-shared/src/hooks/useChainColor.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import { useChainId } from './'
```
Comment (septs) [inline packages/web3-shared/src/hooks/useChainColor.ts:1]:
```suggestion
import { useChainId } from './index'
```

**Comment by wenluomask** [comment]
Forgot the swap in tradeview.

**Comment by zhouhanseng** [comment]
'swap' is out of this PR's region.

**Comment by wenluomask** [comment]
Checked.

**Comment by Jack-Works** [comment]
why close it?

**Review by Jack-Works — APPROVED** [review]

**Review by UncleBill — APPROVED** [review]

**Pull request merged**